### ISSUE TITLE
Not assignable Type

### ISSUE THREAD

**Issue by yuanfux** [microsoft/TypeScript#34967]
**TypeScript 3.7.2**
[Playground link](http://www.typescriptlang.org/play/?ssl=7&ssc=53&pln=7&pc=54#code/MYewdgzgLgBAtgUwhAhgcwQLhtATgSzDRgF4YByACwQBsaQYB3EXGgE3IG4AoUSEGggB09NAApEydAgCUPXuGgwARqXhJUGGAH51UrdgCuYNggBmhBGx58lwNZM0IdepzCMnzl69wWRYKNgA2niEaAA0MMamFmBWALowAD4wQdFecWyRoUSJKSFQBETZhWF5qemxViVFaIlkQcqRwPGcQA)

Compiler Options:
```json
{
  "compilerOptions": {
    "noImplicitAny": true,
    "strictNullChecks": true,
    "strictFunctionTypes": true,
    "strictPropertyInitialization": true,
    "strictBindCallApply": true,
    "noImplicitThis": true,
    "noImplicitReturns": true,
    "useDefineForClassFields": false,
    "alwaysStrict": true,
    "allowUnreachableCode": false,
    "allowUnusedLabels": false,
    "downlevelIteration": false,
    "noEmitHelpers": false,
    "noLib": false,
    "noStrictGenericChecks": false,
    "noUnusedLocals": false,
    "noUnusedParameters": false,
    "esModuleInterop": true,
    "preserveConstEnums": false,
    "removeComments": false,
    "skipLibCheck": false,
    "checkJs": false,
    "allowJs": false,
    "experimentalDecorators": false,
    "emitDecoratorMetadata": false,
    "target": "ES2017",
    "module": "ESNext"
  }
}
```

**Input:**
```typescript
const message: string = 'hello world';
console.log(message);

const b = message ? message : undefined;
const c = message ? message : undefined;

const a: [string, undefined] | [undefined, string] | [string, string] | [undefined, string] = [b, c];
```

**Output:**
```javascript
"use strict";
const message = 'hello world';
console.log(message);
const b = message ? message : undefined;
const c = message ? message : undefined;
const a = [b, c];

```

**Expected behavior:**
To me, Type `[string | undefined, string | undefined]` is assignable to type `[string, undefined] | [undefined, string] | [string, string] | [undefined, undefined]`.

**Comment by AnyhowStep** [comment]
> To me, Type `[string | undefined, string | undefined]` is assignable to type `[string, undefined] | [undefined, string] | [string, string] | [undefined, undefined]`.

Except, it isn't.

```
declare const a : [
  string | undefined, 
  string | undefined
];
//Assume this is okay
const b : (
  | [string, undefined] 
  | [undefined, string] 
  | [string, string] 
  | [undefined, undefined]
) = a;

if (b[0] == undefined) {
  //b is now narrowed to 
  //[undefined, string] | [undefined, undefined] 
  a[0] = "hello, world"; //this is allowed
  console.log(b[0]); //"hello, world"
}
```

**Comment by jack-williams** [comment]
This feels like a bug. As per this [comment](https://github.com/microsoft/TypeScript/issues/30895#issuecomment-482751095), recent changes to discriminant assignment should probably handle this. For example, this works:
```ts
const message: string = 'ntuehon';
console.log(message);

const b = message ? message : undefined;
const c = message ? message : undefined;

interface Pair<L, R> {
    0: L,
    1: R;
}

const a: Pair<string, undefined> | Pair<undefined, undefined> | Pair<string, string> | Pair<undefined, string> = [b, c];
```

**Comment by AnyhowStep** [comment]
Sounds like a bad idea to allow the assignment, since it's unsound

**Comment by jack-williams** [comment]
> I keep telling everyone assignments like these aren't sound and shouldn't be allowed but it's like I'm talking to the wind >.>
And everyone just goes on like I haven't said anything

TypeScript has always allowed covariant references, much like many other languages, which are known to be unsound in the presence of mutation and aliasing.

**Comment by AnyhowStep** [comment]
Except, in this case, it's a contravariant reference, isn't it?

We're trying to assign `string|undefined` to `string`, or `string|undefined` to `undefined`

**Comment by jack-williams** [comment]
It's covariant: `[string | undefined, string | undefined]` is a subtype of `[string, undefined] | [undefined, string] | [string, string] | [undefined, undefined]` because the set of values denoted by the former are a subset of the values denoted by the latter.

Syntax directed subtyping algorithms commonly suffer from the problem that they can't prove this relation using the normal decomposition rules, exactly because you decompose into things that try to related `string | undefined` to `string`.

**Comment by AnyhowStep** [comment]
If mutations are disallowed, I agree with you about the former being a subset of the latter.
If mutations are allowed, then I disagree.

But I guess TS pretends mutations don't happen.

**Comment by RyanCavanaugh** [comment]
@rbuckton n.b. there's a typo (missing the `[undefined, undefined]` case in the OP) but this looks like a case that should have been handled by your PR to look for matching unions

**Comment by rbuckton** [comment]
It looks like this was fixed by #39393, and has been working since 4.0.
